flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamPhysicalMultiJoin.java

```diff
@@ -0,0 +1,221 @@
+
+    /**
+     * A map from input index to a list of {@link ConditionAttributeRef}. Each {@link
+     * ConditionAttributeRef} represents a join condition associated with the input.
+     */
+    private final Map<Integer, List<ConditionAttributeRef>> joinAttributeMap;
+
+    private final @Nullable RexNode postJoinFilter;
+    private final List<RelHint> hints;
+
+    public StreamPhysicalMultiJoin(
+            final RelOptCluster cluster,
+            final RelTraitSet traitSet,
+            final List<RelNode> inputs,
+            final RexNode joinFilter,
+            final RelDataType rowType,
+            final List<? extends @Nullable RexNode> joinConditions,
+            final List<JoinRelType> joinTypes,
+            final Map<Integer, List<ConditionAttributeRef>> joinAttributeMap,
+            final @Nullable RexNode postJoinFilter,
+            final List<RelHint> hints) {
+        super(cluster, traitSet);
+        this.inputs = inputs;
+        this.rowType = rowType;
+        this.joinFilter = joinFilter;
+        this.joinTypes = joinTypes;
+        this.joinConditions = joinConditions;
+        this.joinAttributeMap = joinAttributeMap;
+        this.postJoinFilter = postJoinFilter;
+        this.hints = hints;
+    }
+
+    @Override
+    public boolean requireWatermark() {
+        return false;
+    }
+
+    @Override
+    public List<RelNode> getInputs() {
+        return inputs;
+    }
+
+    @Override
+    public void replaceInput(final int ordinalInParent, final RelNode p) {
+        final List<RelNode> newInputs = new ArrayList<>(inputs);
+        newInputs.set(ordinalInParent, p);
+        this.inputs = List.copyOf(newInputs);
+        recomputeDigest();
+    }
+
+    @Override
+    public RelNode copy(final RelTraitSet traitSet, final List<RelNode> inputs) {
+        return new StreamPhysicalMultiJoin(
+                getCluster(),
+                traitSet,
+                inputs,
+                joinFilter,
+                getRowType(),
+                joinConditions,
+                joinTypes,
+                joinAttributeMap,
+                postJoinFilter,
+                hints);
+    }
+
+    @Override
+    public RelOptCost computeSelfCost(final RelOptPlanner planner, final RelMetadataQuery mq) {
+        final double elementRate = 100.0d * getInputs().size();
+        return planner.getCostFactory().makeCost(elementRate, elementRate, 0);
+    }
+
+    @Override
+    public RelWriter explainTerms(final RelWriter pw) {
+        super.explainTerms(pw);
+        for (final Ord<RelNode> ord : Ord.zip(inputs)) {
+            pw.input("input#" + ord.i, ord.e);
+        }
+        return pw.item("joinFilter", joinFilter)
+                .item("joinTypes", joinTypes)
+                .item("joinConditions", joinConditions)
+                .itemIf("postJoinFilter", postJoinFilter, postJoinFilter != null)
+                .item("select", String.join(",", getRowType().getFieldNames()))
```

Comment:
IIRC usually there should be indexes instead of names
otherwise there might be memory issues if there are too many fields with long names
also I'm not sure the output would be clear for cases where there are several nested subqueries and each time fields have a new alias (for instance set manually in sql)

Reply:
You raise good points. I've also here followed what we did for PTFs https://github.com/apache/flink/blob/86ee4b58e3878335d15aeb425025dd8b05973823/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamPhysicalProcessTableFunction.java#L190

Wdyt @twalthr?